Build tsh with static libfido2 in buildbox and Centos7

codingllama opened the pull request:
Build `tsh` with static `libfido2`, `libcbor`,`libcrypto` and `libudev-zero`.

Dockerfiles for buildbox and Centos7 changed. FIPS and macOS to be addressed at a later date.

Add the `tsh fido2 diag` hidden command for ease of testing.

#9160

codingllama committed:
Add a few build artifacts to .gitignore

codingllama committed:
Update go-libfido2 and tidy modules

codingllama committed:
Add a fido2 diagnostic command to tsh

Useful to test builds.

codingllama commented:
Buildbox:

```diff
$ ldd build/tsh
	linux-vdso.so.1 (0x00007ffd7d989000)
	libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f06c2b5d000)
	libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007f06c2959000)
+ 	libudev.so.1 => /lib/x86_64-linux-gnu/libudev.so.1 (0x00007f06c273b000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f06c234a000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f06c2d7c000)
+ 	librt.so.1 => /lib/x86_64-linux-gnu/librt.so.1 (0x00007f06c2142000)
```

`libudev` and `librt` are readily available, so it's not that bad. [libeudev](https://git.alpinelinux.org/aports/tree/main/libfido2/APKBUILD) could be a reasonable substitution, but we'd need to test it.

Centos:

```diff
$ ldd build/tsh
	linux-vdso.so.1 =>  (0x00007ffcac95c000)
	libpthread.so.0 => /lib64/libpthread.so.0 (0x00007fd76bf19000)
	libdl.so.2 => /lib64/libdl.so.2 (0x00007fd76bd15000)
+ 	libudev.so.1 => /lib64/libudev.so.1 (0x00007fd76baff000)
	libc.so.6 => /lib64/libc.so.6 (0x00007fd76b731000)
	/lib64/ld-linux-x86-64.so.2 (0x00007fd76c135000)
+ 	librt.so.1 => /lib64/librt.so.1 (0x00007fd76b529000)
+ 	libcap.so.2 => /lib64/libcap.so.2 (0x00007fd76b324000)
+ 	libm.so.6 => /lib64/libm.so.6 (0x00007fd76b022000)
+ 	libdw.so.1 => /lib64/libdw.so.1 (0x00007fd76add1000)
+ 	libgcc_s.so.1 => /lib64/libgcc_s.so.1 (0x00007fd76abbb000)
+ 	libattr.so.1 => /lib64/libattr.so.1 (0x00007fd76a9b6000)
+ 	libelf.so.1 => /lib64/libelf.so.1 (0x00007fd76a79e000)
+ 	libz.so.1 => /lib64/libz.so.1 (0x00007fd76a588000)
+ 	liblzma.so.5 => /lib64/liblzma.so.5 (0x00007fd76a362000)
+ 	libbz2.so.1 => /lib64/libbz2.so.1 (0x00007fd76a152000)
```

Centos adds... a lot. Everything seems to be readily available, but it isn't the clean ldd output I'd like to see.

codingllama committed:
Build tsh with static libfido2 in buildbox

codingllama committed:
Build tsh with static libfido2 in centos7

codingllama commented:
FYI @zmb3 @russjones 

Buildbox:

```shell
$ ldd build/tsh
	linux-vdso.so.1 (0x00007ffc5e308000)
	libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007fe89c7de000)
	libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007fe89c5da000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007fe89c1e9000)
	/lib64/ld-linux-x86-64.so.2 (0x00007fe89c9fd000)
```

Centos:

```
$ ldd build/tsh
	linux-vdso.so.1 =>  (0x00007ffdb37e5000)
	libpthread.so.0 => /lib64/libpthread.so.0 (0x00007f06036d6000)
	libdl.so.2 => /lib64/libdl.so.2 (0x00007f06034d2000)
	libc.so.6 => /lib64/libc.so.6 (0x00007f0603104000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f06038f2000)
```

Tried the builds against Debian 7/8 and Ubuntu 14/16/18. Some nice folks tried against Ubuntu 20 as well. The verdict is:

* Debian 7 is a no-go, glibc is too old even for the Centos build
* Debian 8 and Ubuntu 14/16 like the Centos build (glibc again)
* Newer distros are able to run either.

espadolini reviewed (changes requested):
`eudev` is GPL, `libudev` from systemd is LGPL, so we can't use the former and we can only dynamically link the latter. 😭 

Perhaps we can try with [illiliti/libudev-zero](https://github.com/illiliti/libudev-zero)? Either that or we bite the bullet and rewrite libfido2 in go, maybe using [flynn/hid](https://github.com/flynn/hid) for zero-dependency hardware detection.

codingllama commented on build.assets/Dockerfile:
```diff
@@ -6,7 +6,60 @@
+
+# Install libeudev.
+# libeudev replaces systemd's libudev
+RUN git clone --depth=1 https://github.com/eudev-project/eudev.git -b v3.2.11 && \
```
If we are concerned about malicious re-tagging, but tbh I wasn't too concerned about it.

codingllama commented on build.assets/Dockerfile:
```diff
@@ -6,7 +6,60 @@
+
+# Build libfido2 in a separate step to isolate dependencies and give us some
+# flexibility with regards to the base image.
+FROM buildpack-deps:18.04 AS libfido2
```
Yes, pretty much. Stolen from Go's images, but there are a bunch of others that moved to buildpack-deps.

codingllama commented on .gitignore:
```diff
@@ -67,5 +67,11 @@ ssh.config
+/build.assets/.cache
+/build.assets/tooling/bin/**
+
+# Teleport binaries
+/tctl
+/teleport
+/tsh
```
I do a lot of `go build ./tool/tsh`, which spits the binary in the project root, so it's helpful to have the ignores. These are dangerous file names to have anyway, due to `go build`.

codingllama commented on lib/auth/webauthncli/fido2_common.go:
```diff
@@ -16,9 +16,13 @@ package webauthncli
+	"io"
 	"time"
 
+	"github.com/duo-labs/webauthn/protocol"
+	"github.com/duo-labs/webauthn/protocol/webauthncose"
 	"github.com/gravitational/teleport/api/client/proto"
+	"github.com/gravitational/trace"
```
I only tend to go for that in files that already do it, but this package generally doesn't. IMO we need tool-enforced style, import style is all over the place already.

codingllama commented:
> `eudev` is GPL, `libudev` from systemd is LGPL, so we can't use the former and we can only dynamically link the latter. 😭

libfido2 hard-depends on libudev and is BSD-2, which confuses my non-layer brain.  ¯\\_(ツ)_/¯

> Perhaps we can try with [illiliti/libudev-zero](https://github.com/illiliti/libudev-zero)?

I'm OK moving to illiliti/libudev-zero, as it does seem to work. My concern is that, as a single-author library, it's way more likely to be dropped on the floor, whereas eudev appears to be more mature. I do have to say it's a lot easier to build, though...

@zmb3 

> Either that or we bite the bullet and rewrite libfido2 in go, maybe using [flynn/hid](https://github.com/flynn/hid) for zero-dependency hardware detection

Yep, I've had my sights on that for a while. If we can pull off libfido2 I do think that's better, because we leverage Yubico's work. Time is the main factor now, as I can't make the next release if I have to rewrite libfido2, but post release this could change.

codingllama committed:
Add a few relevant cmake flags

codingllama committed:
Use illiliti/libudev-zero

codingllama committed:
Do multi-stage build on centos7, image tweaks

codingllama committed:
Add `make enter/centos7`

Convenience target. The centos7 image is base for a few others, which
makes it important.

codingllama commented:
PTAL? Now building with illiliti/libudev-zero.

espadolini commented on build.assets/Dockerfile:
```diff
@@ -6,7 +6,44 @@
+RUN git clone --depth=1 https://github.com/PJK/libcbor.git -b v0.9.0 && \
+    cd libcbor && \
+    cmake \
+        -DCBOR_CUSTOM_ALLOC=ON \
```
Have we figured out what this is for?

codingllama replied:
It seems to be their preferred setting, for example see the README: https://github.com/PJK/libcbor#compile-from-source. There are [mentions in issues](https://github.com/PJK/libcbor/issues/147#issuecomment-666401486) about it being unfortunate that it is disabled by default. I don't know much more than that, though.

espadolini commented on build.assets/Dockerfile:
```diff
@@ -6,7 +6,44 @@
+    CFLAGS=-pthread cmake \
+        -DBUILD_EXAMPLES=OFF \
+        -DBUILD_MANPAGES=OFF \
+        -DBUILD_TOOLS=OFf \
```
```suggestion
        -DBUILD_TOOLS=OFF \
```

codingllama replied:
Done.

espadolini commented on build.assets/Dockerfile:
```diff
@@ -188,6 +207,21 @@ RUN curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh -s -- -y --pr
+COPY --from=libfido2 \
+    /usr/local/lib/libcbor.a \
+    /usr/local/lib/libfido2.a \
+    /usr/local/lib/libfido2.so.1.10.0 \
```
I would avoid building/copying the shared library altogether, to be sure we don't accidentally link it. (same applies to the centos build)

espadolini replied:
Wouldn't that fail to test problems that might occur as a result of using the statically linked library?

codingllama replied:
Testing mocks fido2 interactions, so unfortunately no. We do need the libraries to `go get` and compile go-libfido2.

It would be interesting to exercise and `ldd` our release binaries, to make sure we have no regressions, but that must happen outside of the scope of the Dockerfile (they don't actually capture the sources).

espadolini commented on build.assets/Dockerfile-centos7:
```diff
@@ -1,4 +1,65 @@
+    cmake3 \
+        -DBUILD_EXAMPLES=OFF \
+        -DBUILD_MANPAGES=OFF \
+        -DBUILD_TOOLS=OFf \
```
```suggestion
        -DBUILD_TOOLS=OFF \
```

codingllama replied:
Done.

codingllama committed:
s/OFf/OFF/g

codingllama commented:
Friendly ping @timothyb89.

timothyb89 commented on Makefile:
```diff
@@ -156,10 +156,22 @@ endif
+# system. The required libs are uncommon enough that their presence makes for a
+# strong signal.
+LIBFIDO2_MESSAGE := without libfido2
+ifneq ("$(wildcard /usr/local/lib/libfido2.a)", "")
+ifneq ("$(wildcard /usr/local/lib/libcbor.a)", "")
+ifneq ("$(wildcard /usr/local/lib/libudev.a)", "")
```
Maybe out of scope for this PR but it's a bit unfortunate we hard-code `/usr/local/lib` which isn't very sensible on multiarch systems. I'm not sure I have a better solution to offer though since libcbor doesn't seem to play with pkg-config or anything.

codingllama replied:
I'm not in love this, but at least it gets us out of the door. There's definitely room for improvement.

espadolini replied:
FWIW I get a pkg-config entry for libcbor and libfido2 when installing them on macOS through brew

codingllama replied:
👍 

I was taking a look at pkgconfig over the weekend, I'll see if I carve some time to give it a try, specially for the static builds.
